modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/template/HeadlessHttpClientTemplateContextContributor.java

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.vulcan.internal.template;
+
+import com.liferay.petra.io.unsync.UnsyncStringWriter;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.json.JSONFactoryUtil;
+import com.liferay.portal.kernel.servlet.PipingServletResponse;
+import com.liferay.portal.kernel.servlet.ServletContextPool;
+import com.liferay.portal.kernel.template.TemplateContextContributor;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.vulcan.internal.template.servlet.HeadlessHttpClientHttpRequestWrapper;
+import com.liferay.portal.vulcan.internal.template.servlet.HeadlessHttpClientHttpResponseWrapper;
+
+import java.util.Map;
+
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletContext;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * @author Alejandro Tardín
+ */
+@Component(
+	property = "type=" + TemplateContextContributor.TYPE_GLOBAL,
+	service = TemplateContextContributor.class
+)
+public class HeadlessHttpClientTemplateContextContributor
+	implements TemplateContextContributor {
+
+	@Override
+	public void prepare(
+		Map<String, Object> contextObjects,
+		HttpServletRequest httpServletRequest) {
+
+		ThemeDisplay themeDisplay = (ThemeDisplay)contextObjects.get(
+			"themeDisplay");
+
+		contextObjects.put(
+			"headlessHttpClient",
+			new HeadlessHttpClient(
+				httpServletRequest, themeDisplay.getResponse()));
+	}
+
+	public class HeadlessHttpClient {
+
+		public HeadlessHttpClient(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse) {
+
+			_httpServletRequest = httpServletRequest;
+			_httpServletResponse = httpServletResponse;
+		}
+
+		public Object get(String path) throws Exception {
+			UnsyncStringWriter unsyncStringWriter = new UnsyncStringWriter();
+
+			ServletContext servletContext = ServletContextPool.get(
```

Comment:
@4lejandrito it looks fine overall. Just since you always just get the RequestDispatcher from portal servlet context, you should hold the portal ServletContext as a field inside HeadlessHttpClientTemplateContextContributor/HeadlessHttpClient to avoid runtime lookup on every RequestDispatcher getting.

Reply:
Sounds good I’ll do that. Thanks!!

@brianchandotcom.

Reply:
See https://github.com/liferay-headless/liferay-portal/pull/819/commits/65d53f5fe8512d9986190338120869ffa83c61e7.